src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -501,6 +501,18 @@ public string TraceStateString
                         _traceId = ActivityTraceId.CreateFromString(_id.AsSpan(3, 32));
                         _traceIdSet = true;
                     }
+                    else if (_parentId != null && IsW3CId(_parentId))
+                    {
+                        try
+                        {
+                            _traceId = ActivityTraceId.CreateFromString(_parentId.AsSpan(3, 32));
+                            _traceIdSet = true;
+                        }
+                        catch
+                        {
+                            // ActivityTraceId.CreateFromString throws if parent is invalid W3C Id
+                        }
+                    }
```

Comment:
I don't like the fact that we are cloning this code, and the code in GenerateW3CId().   

Here is my suggestion:

Take all the code in the in the "if (!_traceIdSet)" clause in GenerateW3Id and put it in here (I think it can replace almost all of the code in this method.

Then call this TraceId property from GenerateW3CId() (which now just has this call and l.   Now we have only one copy (that lives in the TraceId property).   If you don't like the property that you call just for the side effect, it is OK by me to make a helper that is called both places.